roomserver/internal/perform/perform_upgrade.go

```diff
@@ -0,0 +1,691 @@
+		util.GetLogger(ctx).WithError(err).Error()
+		return nil, &api.PerformError{
+			Msg: "Membership event content was invalid",
+		}
+	}
+	membershipContent.Membership = gomatrixserverlib.Join
+	membershipEvent := fledglingEvent{
+		Type:     gomatrixserverlib.MRoomMember,
+		StateKey: userID,
+		Content:  membershipContent,
+	}
+
+	powerLevelContent, err := oldPowerLevelsEvent.PowerLevels()
+	if err != nil {
+		util.GetLogger(ctx).WithError(err).Error()
+		return nil, &api.PerformError{
+			Msg: "Power level event content was invalid",
+		}
+	}
+
+	// We might need to temporarily give ourselves a higher power level
+	// than we had in the old room in order to be able to send all of
+	// the relevant state events. This function will return whether we
+	// had to override the power level events or not — if we did, we
+	// need to send the original power levels again later on.
+	tempPowerLevelsEvent, powerLevelsOverridden := createTemporaryPowerLevels(powerLevelContent, userID)
+
+	joinRulesContent, err := oldJoinRulesEvent.JoinRule()
+	if err != nil {
+		return nil, &api.PerformError{
+			Msg: "Join rules event content was invalid",
+		}
+	}
+	newJoinRulesEvent := fledglingEvent{
+		Type: gomatrixserverlib.MRoomJoinRules,
+		Content: map[string]interface{}{
+			"join_rule": joinRulesContent,
+		},
+	}
+
+	eventsToMake := make([]fledglingEvent, 0, len(state))
+	eventsToMake = append(eventsToMake, newCreateEvent, membershipEvent, tempPowerLevelsEvent, newJoinRulesEvent)
+
+	if _, ok := state[gomatrixserverlib.StateKeyTuple{EventType: gomatrixserverlib.MRoomGuestAccess, StateKey: ""}]; !ok {
+		// Appease sytest, as it expects a guest access event for some reason.
+		eventsToMake = append(eventsToMake, fledglingEvent{
+			Type: gomatrixserverlib.MRoomGuestAccess,
+			Content: map[string]string{
+				"guest_access": "forbidden",
+			},
+		})
+	}
+
+	// Duplicate all of the old state events into the new room.
+	for tuple, event := range state {
+		if _, ok := override[tuple]; ok {
+			// Don't duplicate events we have overridden already. They
+			// are already in `eventsToMake`.
+			continue
+		}
+		if membership, merr := event.Membership(); merr == nil && membership != gomatrixserverlib.Ban {
+			// Don't duplicate membership events that aren't bans. Our own
+			// membership event has already been created above, and event auth
+			// won't let us create join membership events for other users.
+			continue
+		}
+		newEvent := fledglingEvent{
+			Type:     tuple.EventType,
+			StateKey: tuple.StateKey,
+		}
+		if err = json.Unmarshal(event.Content(), &newEvent.Content); err != nil {
+			logrus.WithError(err).Error("Failed to unmarshal old event")
+			continue
+		}
+		eventsToMake = append(eventsToMake, newEvent)
+	}
+
+	// If we sent a temporary power level event into the room before,
+	// override that now by restoring the original power levels.
+	if powerLevelsOverridden {
```

Comment:
I don't understand why this is necessary. We are _creating the room_ ergo we will have the power to do anything we like. Why aren't we just making the room, sending all the state bar the PL event, then send the PL event (which may drop our privileges?)

Reply:
I think because of this: https://github.com/matrix-org/gomatrixserverlib/blob/0980b7f341e0c70adc2784eb8a978578835bbd85/eventauth.go#L266-L271